client/src/components/requestPanel/RequestPanel.tsx

```diff
@@ -20,34 +20,74 @@ import KVEditor from '../kvEditor';
 import UriBar from '../uriBar';
 import styles from './RequestPanel.module.css';
 
-const defaultParam = {
-  key: '',
-  value: '',
-};
+function getParamsFromUri(uri: string, params?: Array<KVRow>): Array<KVRow> {
+  const paramString = uri.split('?')[1];
+  if (!paramString) {
+    return params?.filter((param) => param.isEnabled === false) ?? [];
+  }
+
+  const uriParams = paramString.split('&').map((kv) => {
+    const [k, ...v] = kv.split('='); // ...v with v.join('=') handle cases where the value contains '='
+    return {
+      key: k,
+      value: v.join('='),
+    };
+  });
+
+  if (!params) {
+    return uriParams;
+  }
 
-function shouldAppendNewRow(params: Array<KVRow>): boolean {
-  if (params.length === 0) return true;
-  const { key, value } = params[params.length - 1];
-  return key !== '' || value !== '';
+  const newParams: KVRow[] = [];
+
+  let indexEnabledParams = 0;
+  for (const [_, param] of params.entries()) {
+    if (param.isEnabled === false) {
+      newParams.push(param);
+    } else {
+      const uriParam = uriParams[indexEnabledParams];
+      if (!uriParam) {
+        console.error('params and URI params out of sync');
+        newParams.push({ key: '', value: '' });
+      } else {
+        newParams.push(uriParam);
+      }
+      indexEnabledParams++;
+    }
+    if (uriParams[indexEnabledParams]) {
```

Comment:
also this block is not needed

Reply:
Why don't we need this? This is a signal to notify us that we have gone out of sync, in situations where uriParams > params.
We can make it a warning if that's better

Reply:
but you already log on top `console.error('params and URI params out of sync');`

It's the same case you are checking.

Reply:
No, the top case and the bottom case are different. 
Check the latest commit, it's a bit more clear there.
The error case in the top log does only apply for `enabled Params > uriParams` not apply for `uriParams > enabled Params`, so we wouldn't notice something went wrong in the logs.

Reply:
OK. Then we can move the error log in the below for loop where we append the missing rows. Having it as `if (uriParams[indexEnabledParams])` takes a lot more understanding of the rest of the code.

 Rest LGTM.

Reply:
Made it more understandable in the latest commit